Source/x11/XGServerEvent.m

```diff
@@ -1945,15 +1958,36 @@ - (NSEvent *)_handleTakeFocusAtom: (XEvent)xEvent
          window to take focus after each one gets hidden. */
       NSDebugLLog(@"Focus", @"WM take focus while hiding");
     }
-  else if (cWin->ignore_take_focus == YES)
+  else if ([[NSApp mainMenu] isTransient] != NO) // right-click main menu
+    {
+      /* Do not grab focus from active application if right-click on our 
+         application icon was performed. */
+      NSDebugLLog(@"Focus",
+                  @"[TakeFocus] ignore request for transient application menu.");
+    }
+  else if (cWin->number == key_num && keyWin->map_state != IsUnmapped) // already key window
     {
       NSDebugLLog(@"Focus", @"Ignoring window focus request");
       cWin->ignore_take_focus = NO;
     }
-  else if (cWin->number == key_num)
+  else if (cWin->ignore_take_focus == YES) // after orderwindow:::
     {
-      NSDebugLLog(@"Focus", @"Reasserting key window");
-      [GSServerForWindow(key_win) setinputfocus: key_num];
+      NSDebugLLog(@"Focus",
+                  @"[TakeFocus] %lu(%li): ignore_take_focus == YES...",
+                  cWin->ident, cWin->number);
+      /* Window was requested to take focus and earlier (orderwindow:::) 
```

Comment:
You seem to be implying that key_win and cWin are the same window. But why do you think so?

Reply:
I don't see `key_win` in the cited code? Do you mean code across method? If so, my answer is: I'm not implying this, I'm checking if this is a true at some points. Or you meant some other part of code?

Reply:
Just a few lines below 1978, there you check the map_state of keyWin but then set the focus on cWin. In the line 1983 you even get the server from key_win. I really don't understand when this is needed, but the code looks wrong.
Maybe if we could go through the different cases in this method one by one?
The annoying bit about this code here is that all the existing special cases where written because they seemed necessary for a specific window manager. Now you are throwing some of the special stuff away to add your own. The will surely work on your specific WindowMaker setup, but what about other systems?

Reply:
`keyWin` and `key_win` represents the same window but different structures: gs_window_device_t and NSWindow respectively. It is a key window as it sees NSApplication. `cWin` is a current focused window (look NSDebugLLog message at line 1922, this is not my text). So code you've pointed me at checks if NSApplication and X11 vision of what the focus window is and performs respective actions. What's wrong with that?

Other window managers are out of my interest. I expect other developers/users try it and give me feedback. Otherwise - no problem, I'm able to maintain this patch in NEXTSPACE.

Reply:
You're right actually. I'll check this part of code. Do you have comments to other parts of this patch?